### ISSUE TITLE
modification on Allgather: how to achieve a random sampling process

### ISSUE THREAD

**Issue by zrsjta** [horovod/horovod#1583]
Hello, dear researchers.
In recent work, I try to implement my own parallel training algorithm: for each worker, instead of gathering from all ranks, only the data of tensors from several neighbor ranks would be sampled. The key feature of this algorithm is that, for different training step / different variable / different rank, the sampling target should randomly change. The details of the algorithm is shown in this paper : 
[https://arxiv.org/abs/1909.09010](url)

To achieve this algorithm, I force the Horovod works only with MPI and make a modification mainly in function: `/common/ops/mpi_operation.cc MPIAllgather::Execute.`  Roughly speaking, for each time the `execute` function is called, I generate two array:` local_sends` and `local_recvs` to indicate the communication targets on this rank. Then I replace the MPI_Allgatherv by MPI_Sendrecv, MPI_Send and MPI_Recv to complete the communication. 
The code of this modification is shown as follow. Note that the generation process is in function 
`GenerateSamplingList` and `GenerateCommSeq`, and I use code `std::srand((unsigned)time(NULL));` 
to ensure a different random seed

```
Status MPIAllgather::Execute(std::vector<TensorTableEntry>& entries, const Response& response) {
  auto& timeline = global_state_->timeline;
  auto** entry_component_sizes = new int64_t* [entries.size()]; 
  auto** entry_component_offsets = new int64_t* [entries.size()];

  int global_size = global_state_->controller->GetSize();
  int global_rank = global_state_->controller->GetRank();
  
  auto* sampling_list = new int[sample_size * global_size]();
  if (global_rank == 0){GenerateSamplingList(global_size, sample_margin, sample_size, sampling_list);} 
  MPI_Bcast(sampling_list, sample_size * global_size, MPI_INT, 0, mpi_context_->GetMPICommunicator(Communicator::GLOBAL));
  
  auto* global_dests = new int[global_size * sample_margin * max_round_factor]();
  auto* global_srcs = new int[global_size * sample_margin * max_round_factor]();
  int global_comm_rounds = 0;
  GenerateCommSeq(sampling_list, global_size, global_rank, max_round_factor, sample_size, sample_margin, 
      global_dests, global_srcs, global_comm_rounds);
  
  auto* local_recvs = new int[global_comm_rounds]();
  auto* local_sends = new int[global_comm_rounds]();
  for (int i = 0; i < global_comm_rounds; i++){
      local_recvs[i] = global_dests[i * global_size + global_rank]; 
      local_sends[i] = -1;
      for (int j = 0; j < global_size; j++){
        if(global_dests[i * global_size + j] == global_rank) {local_sends[i] = j;}
      }
  }
  // here we finish logics of gossip src / target topology

  auto* recvcounts = new int[sample_size]();
  auto* displcmnts = new int[sample_size]();

  for (size_t ec = 0; ec < entries.size(); ++ec) {
    entry_component_sizes[ec] = new int64_t[sample_size]();
    entry_component_offsets[ec] = new int64_t[sample_size]();
  }
  // prepare recvcounts and displacements for rank and components
  Status status = AllocateOutput(entries, response, local_recvs, global_comm_rounds, entry_component_sizes, recvcounts);
  SetDisplacements(recvcounts, sample_size, displcmnts);
  SetEntryComponentOffsets(entries, entry_component_sizes, recvcounts, sample_size, entry_component_offsets);
  // prepare send and recv size / element size
  int64_t num_elements_send = NumElements(entries);
  int64_t num_elements_recv = 0; 
  for (int i = 0; i < sample_size; i++){num_elements_recv+=recvcounts[i];}
  auto& first_entry = entries[0];
  int element_size = mpi_context_->GetMPITypeSize(first_entry.tensor->dtype());
  auto dtype = mpi_context_->GetMPIDataType(first_entry.tensor->dtype());
  MPI_Status mpi_status;

  const void* send_buf = nullptr; // send buffer
  void* recv_buf;  // recv buffer
  void* fetch_buf; // if use fusion buffer, the recv buffer would shift
  
  if (entries.size() > 1) {
    // recv_buf is the starting point of the fusion buffer
    // send_buf is shifted by num_elements_recv units 
    timeline.ActivityStartAll(entries, MEMCPY_IN_FUSION_BUFFER);
    MemcpyInFusionBuffer(entries, displcmnts, element_size, num_elements_recv, recv_buf);
    send_buf = recv_buf + num_elements_recv * element_size;
    fetch_buf = recv_buf; 
    timeline.ActivityEndAll(entries);
  } else {
    send_buf = first_entry.tensor->data();
    recv_buf = (void*) first_entry.output->data();
  }

  int recv_idx = 0;
  for (int round = 0; round < global_comm_rounds; round++){
    if ((local_sends[round] != -1) && (local_recvs[round] != -1)){ 
      MPI_Sendrecv(send_buf,      
                (int)num_elements_send,
                dtype,
                local_sends[round],
                round * global_size + global_rank,
                recv_buf,  
                (int)recvcounts[recv_idx],
                dtype,
                local_recvs[round],
                round * global_size + local_recvs[round],
                mpi_context_->GetMPICommunicator(Communicator::GLOBAL),  
                &mpi_status
                );
      recv_buf = recv_buf + recvcounts[recv_idx] * element_size;
      ++ recv_idx ;
      }
    
    else if ((local_sends[round] != -1) && (local_recvs[round] == -1)){ 
      MPI_Send(send_buf,
              (int)num_elements_send,
              dtype,
              local_sends[round],
              round * global_size + global_rank,
              mpi_context_->GetMPICommunicator(Communicator::GLOBAL));
    }
    
    else if ((local_sends[round] == -1) && (local_recvs[round] != -1)){
      MPI_Recv(recv_buf,  
              (int)recvcounts[recv_idx],
              dtype,
              local_recvs[round],
              round * global_size + local_recvs[round],
              mpi_context_->GetMPICommunicator(Communicator::GLOBAL),
              &mpi_status);
      recv_buf = recv_buf + recvcounts[recv_idx] * element_size;
      ++recv_idx ;

    }
    else if ((local_sends[round] == -1) && (local_recvs[round] == -1)){}
    auto MPI_Barrier(mpi_context_->GetMPICommunicator(Communicator::GLOBAL));
  }
  
  global_state_->timeline.ActivityEndAll(entries);

  if (entries.size() > 1) {
    timeline.ActivityStartAll(entries, MEMCPY_OUT_FUSION_BUFFER);
    MemcpyOutFusionBuffer(entry_component_offsets, entry_component_sizes,
                          fetch_buf, element_size, sample_size, entries);
    timeline.ActivityEndAll(entries);
  }

  delete[] recvcounts;
  delete[] displcmnts;

  for (size_t ec = 0; ec < entries.size(); ++ec) {
    delete[] entry_component_sizes[ec];
    delete[] entry_component_offsets[ec];
  }
  delete[] entry_component_sizes;
  delete[] entry_component_offsets;

  delete[] sampling_list;

  delete[] local_recvs;
  delete[] local_sends;

  delete[] global_dests;
  delete[] global_srcs;


  return Status::OK();
}
```

and I test this function by python scripts as follow:
`mpirun -np 8 -bind-to none -map-by slot -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH -mca pml ob1 -mca btl ^openib python -u show_problem.py`

```
import tensorflow as tf
import horovod.tensorflow as hvd
import os
from mpi4py import MPI
import time

comm = MPI.COMM_WORLD
os.environ["CUDA_VISIBLE_DEVICES"]=""
os.environ["HOROVOD_GOSSIP_SIZE"]="1"
os.environ["HOROVOD_GOSSIP_MARGIN"]="2"
os.environ["HOROVOD_GOSSIP_MAX_ROUND_FACTOR"]="8"

hvd.init()
size = hvd.size()
rank = hvd.rank()
local_rank = hvd.local_rank()

a = tf.ones([1]) * (rank)
b = tf.ones([rank+1]) * (rank)
c = hvd.allgather(a)
time.sleep(1)
d = hvd.allgather(c)

e = tf.range(5)
f = tf.random.shuffle(e)
cpu_name = "/cpu:{}".format(local_rank)
comm.Barrier()
with tf.device(cpu_name):
    with tf.Session() as sess:
        for i in range(5):
            c_ = sess.run(c)
            time.sleep(1)
            d_ = sess.run(d)
            if rank == 0:
                print(" iter: {} rank : {} after gossip a is {}".format(i, rank, c_))
                print(" iter: {} rank : {} after gossip c is {}".format(i, rank, d_))
            comm.Barrier()
        if rank == 0:
            print("so the our sampling target is deterministic in each call")

        for i in range(5):
            f_ = sess.run(f)
            if rank == 0 :
                print(" iter: {} after tensorflow random shuffle e is {}".format(i, f_))
                if i == 4:
                    print(" so the tensorflow random shuffle is random in each call")
```
the results is as follow:

```
iter: 0 rank : 0 after gossip a is [6.]
 iter: 0 rank : 0 after gossip c is [3.]
 iter: 1 rank : 0 after gossip a is [6.]
 iter: 1 rank : 0 after gossip c is [3.]
 iter: 2 rank : 0 after gossip a is [6.]
 iter: 2 rank : 0 after gossip c is [3.]
 iter: 3 rank : 0 after gossip a is [6.]
 iter: 3 rank : 0 after gossip c is [3.]
 iter: 4 rank : 0 after gossip a is [6.]
 iter: 4 rank : 0 after gossip c is [3.]
so the our sampling target is deterministic in each call
 iter: 0 after tensorflow random shuffle e is [3 0 2 4 1]
 iter: 1 after tensorflow random shuffle e is [3 2 1 0 4]
 iter: 2 after tensorflow random shuffle e is [3 1 0 2 4]
 iter: 3 after tensorflow random shuffle e is [0 1 4 3 2]
 iter: 4 after tensorflow random shuffle e is [0 1 3 4 2]
 so the tensorflow random shuffle is random in each call
```

It seems that, the sampling targets for different training steps are the same.
In my view, each time the `allgather` function is called in tensorflow, the execute function should be conducted once, and then the` local_sends`, `local_recvs ` should be newly generated, and it should keep changing in different steps, just like how` tf.random.shuffle` behaves: different results for different calls
Also, when I print the `local_sends` `local_recvs` in the execute function, they only appear once and maybe they have been recorded in tensorflow graph somehow, or due to some other reasons.

So I want to ask : would the execute function be called for each time tensorflow calles the allgather operation?  How can I fix this problem to obtain different` local_sends` `local_recvs` values in different training steps?  

Note: several code files have been attached in this issue in txt form, please check if needed.

Thanks for your help! 

[collective_operation_cc.txt](https://github.com/horovod/horovod/files/3959693/collective_operation_cc.txt)
[collective_operation_h.txt](https://github.com/horovod/horovod/files/3959694/collective_operation_h.txt)
[mpi_operation_cc.txt](https://github.com/horovod/horovod/files/3959695/mpi_operation_cc.txt)
[mpi_operation_h.txt](https://github.com/horovod/horovod/files/3959696/mpi_operation_h.txt)

**Comment by tgaddair** [comment]
Hey @zrsjta, this is an interesting approach. Do you think you would be interested in turning this into a PR at some point?

To answer your question about `Execute`, it is not necessarily the case that one call to `allgather` from TensorFlow maps to one call to `Execute` in Horovod, due to tensor fusion.  But I think what you mean is: should `Execute` be called every time I call `sess.run()`?

In this case, `Execute` is only being called once because TensorFlow is doing something called [constant folding](https://stackoverflow.com/questions/39198213/do-subsequent-calls-to-sess-run-erase-state) which caches the return value of `hvd.allgather`, because the inputs haven't changed between calls to `sess.run()`.

One way to get around this is to use `tf.Variable` for your inputs:

```
a = tf.ones([1]) * (rank)
a2 = tf.Variable(a)
c = hvd.allgather(a2)

with tf.Session() as sess:
     sess.run(tf.global_variables_initializer())
     for i in range(5):
         c_ = sess.run(c)
         if rank == 0:
             print(" iter: {} rank : {} after gossip c is {}".format(i, rank, c_))
```

A more robust solution would be to implement this at the op level in Horovod.  If you compare Horovod custom ops with TensorFlow's [random_ops.cc](https://github.com/tensorflow/tensorflow/blob/v1.11.0/tensorflow/core/ops/random_ops.cc) you'll notice that they get around this caching mechanism by declaring their ops with `SetIsStateful()`.

I would also recommend reading [this](https://stackoverflow.com/questions/52600729/how-to-use-random-number-in-user-defined-tensorflow-op) StackOverflow post about how TensorFlow ops implement randomness, which covers a lot of good information as well.

**Comment by zrsjta** [comment]
@tgaddair Thanks! the problem has been fixed, and this is a feedback

1. The idea of using` tf.Variable ` still cannot work for me. It seems that TensorFlow would cache the output of this operation as long as the value of the variables stay the same, no matter it is a constant or a variable. Here is an example.
`mpirun -np 8 -bind-to none -map-by slot -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH -mca pml ob1 -mca btl ^openib python -u test.py`

```
import tensorflow as tf
import horovod.tensorflow as hvd
import os
from mpi4py import MPI
import time


comm = MPI.COMM_WORLD
os.environ["CUDA_VISIBLE_DEVICES"]=""
os.environ["HOROVOD_GOSSIP_SIZE"]="1"
os.environ["HOROVOD_GOSSIP_MARGIN"]="2"
os.environ["HOROVOD_GOSSIP_MAX_ROUND_FACTOR"]="8"

hvd.init()
size = hvd.size()
rank = hvd.rank()
local_rank = hvd.local_rank()

a = tf.ones([1]) * (rank)
a2 = tf.Variable(a)
c = hvd.allgather(a)

cpu_name = "/cpu:{}".format(local_rank)
comm.Barrier()
with tf.device(cpu_name):
    with tf.Session() as sess:
        for i in range(5):
            c_ = sess.run(c)
            if rank == 0:
                print(" iter: {} rank : {} after gossip a is {}".format(i, rank, c_))
            comm.Barrier()

```
output :
```
iter: 0 rank : 0 after gossip a is [1.]
 iter: 1 rank : 0 after gossip a is [1.]
 iter: 2 rank : 0 after gossip a is [1.]
 iter: 3 rank : 0 after gossip a is [1.]
 iter: 4 rank : 0 after gossip a is [1.]
```
But this can work in my real project as gradients are applied repetitively, and the `execute`  function, as checked, would be called in each training step. So the problem has been fixed.

2.  In previous work, I use the Horovod allgather operation, and then shuffle the results and slice them to implement the desired components on other ranks. It seems the current modification run faster than the previous version if only one neighbor rank is sampled by each rank (with 4 GPUs). But currently I haven't test it with other settings.

3.  further questions are here: would the broadcast operation in the `execute` function significantly harm the running speed? I mean, is it very necessary to merge this process in the Request-Response mechanism. Also, is there any reference material about TensorFlow underlying architecture or custom operation developing to share :) Finally, dose PR means "pull request", which means a share of my code? I am glad to do so but still not familiar with git and github. I would learn it later and share the code if Horovod needs that.

**Comment by tgaddair** [comment]
Hey @zrsjta, there may be an issue with the way you're using `tf.Variable`.  When I ran a similar test, I got different results for each call to `hvd.allgather()`.  Feel free to post your code and I'd be happy to test it out.

A broadcast operation would slow things down, but how much would vary a lot in different situations.  In your case, because you're using a sampling process, you should be able to reduce the total amount of data sent over the network in exchange for more requests, which could be a net positive in bandwidth constrained environments.  We have have some docs on contributing to Horovod [here](https://github.com/horovod/horovod/blob/master/docs/contributors.rst).  Mostly you wouldn't be dealing with TensorFlow custom ops.

Yes, PR is short for "pull request".  No pressure to do so if it's too much trouble, but contributing to an open source project can look good on a resume :).
